components/ai_chat/core/browser/engine/conversation_api_client.cc

```diff
@@ -0,0 +1,299 @@
+        event_dict.Set("type", "requestSummary");
+        break;
+      case ConversationEventType::RequestSuggestedActions:
+        event_dict.Set("type", "requestSuggestedActions");
+        break;
+      case ConversationEventType::SuggestedActions:
+        event_dict.Set("type", "suggestedActions");
+        break;
+      default:
+        NOTREACHED();
+        break;
+    }
+    event_dict.Set("content", event.content);
+    events.Append(std::move(event_dict));
+  }
+  return events;
+}
+
+GURL GetEndpointUrl(bool premium, const std::string& path) {
+  DCHECK(!path.starts_with("/"));
+
+  auto* prefix = premium ? "ai-chat-premium.bsg" : "ai-chat.bsg";
+  auto hostname = brave_domains::GetServicesDomain(
+      prefix, brave_domains::ServicesEnvironment::DEV);
+
+  GURL url{base::StrCat(
+      {url::kHttpsScheme, url::kStandardSchemeSeparator, hostname, "/", path})};
+
+  DCHECK(url.is_valid()) << "Invalid API Url: " << url.spec();
+
+  return url;
+}
+
+}  // namespace
+
+ConversationAPIClient::ConversationAPIClient(
+    const std::string& model_name,
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory,
+    AIChatCredentialManager* credential_manager)
+    : model_name_(model_name),
+      api_request_helper_(GetNetworkTrafficAnnotationTag(), url_loader_factory),
+      credential_manager_(credential_manager) {
+  DCHECK(!model_name_.empty());
+}
+
+ConversationAPIClient::~ConversationAPIClient() = default;
+
+void ConversationAPIClient::ClearAllQueries() {
+  api_request_helper_.CancelAll();
+}
+
+void ConversationAPIClient::PerformRequest(
+    const std::vector<ConversationEvent>& conversation,
+    GenerationDataCallback data_received_callback,
+    GenerationCompletedCallback completed_callback) {
+  // Get credentials and then perform request
+  auto callback = base::BindOnce(
+      &ConversationAPIClient::PerformRequestWithCredentials,
+      weak_ptr_factory_.GetWeakPtr(), std::move(conversation),
+      std::move(data_received_callback), std::move(completed_callback));
+  credential_manager_->FetchPremiumCredential(std::move(callback));
+}
+
+std::string ConversationAPIClient::CreateJSONRequestBody(
+    const std::vector<ConversationEvent>& conversation,
+    const bool is_sse_enabled) {
+  DCHECK(!model_name_.empty());
+  base::Value::Dict dict;
+
+  dict.Set("events", ConversationEventsToList(conversation));
+  dict.Set("model", model_name_);
+  dict.Set("stream", is_sse_enabled);
+
+  std::string json;
+  base::JSONWriter::Write(dict, &json);
+  return json;
+}
+
+void ConversationAPIClient::PerformRequestWithCredentials(
+    const std::vector<ConversationEvent>& conversation,
+    GenerationDataCallback data_received_callback,
+    GenerationCompletedCallback completed_callback,
+    std::optional<CredentialCacheEntry> credential) {
+  bool premium_enabled = credential.has_value();
+  const GURL api_url = GetEndpointUrl(premium_enabled, kRemotePath);
```

Comment:
It seems a little unnecessary since these are all static values (the host and the path), and seems more for a CHECK, and IMO one inside `GetEndpointUrl` but that's debatable I suppose. Happy to return early here too.

Reply:
We'd need a UI error state like "Unknown Error" and it would be impossible to have that show unless something changes in a future development.

Reply:
See https://github.com/brave/brave-core/pull/23067/files/65cc51401f63413ce57caca68eb7ffb387a511c5#r1591374647 for this particular case.
I still think in the future it would be nice to have an internal error state in mojom::APIError that backend could use during error handling, but does not need to be in this PR.

Reply:
I ended up doing this and just using `APIError:None`